File: 2516-take-k-of-each-character-from-left-and-right/2516-take-k-of-each-character-from-left-and-right.cpp
using vi = vector<int>;
using vvi = vector<vi>;

class Solution {
public:
  int takeCharacters(string s, int k) {
    int n = s.size();
    vvi pre(n+1, vi(3));
    for (int i=0; i<n; i++) {
      pre[i+1] = pre[i];
      pre[i+1][s[i]-'a']++;
    }

    int ans = -1;
    int lo = 0, hi = n;
    while (lo <= hi) {
      int mid = (lo+hi) / 2;
      bool ok = false;
      for (int i=0; i<=mid; i++) {
        bool flag = true;
        for (int j=0; j<3; j++) {
          flag &= (pre[i][j] + pre[n][j] - pre[n-mid+i][j] >= k);
        }
        if (flag) {
          ok = true;
          break;
        }
      }
      if (ok) {
        ans = mid;
        hi = mid - 1;
      }
      else lo = mid + 1;
    }
    return ans;
  }
};

File: 2516-take-k-of-each-character-from-left-and-right/2516-take-k-of-each-character-from-left-and-right.py
class Solution:
    def takeCharacters(self, s: str, k: int) -> int:
        n = len(s)
        pre = [[0] * 3 for _ in range(n + 1)]

        for i, c in enumerate(s):
            pre[i + 1][0], pre[i + 1][1], pre[i + 1][2] = pre[i][0], pre[i][1], pre[i][2]
            pre[i + 1][ord(c) - ord('a')] += 1

        ans = -1
        lo, hi = 0, n

        while lo <= hi:
            mid = (lo+hi) // 2
            ok = False
            for i in range(mid+1):
                flag = True
                for j in range(3):
                    flag &= pre[i][j] + pre[n][j] - pre[n-mid+i][j] >= k
                if flag:
                    ok = True
                    break
            if ok:
                ans = mid
                hi = mid - 1
            else:
                lo = mid + 1
        
        return ans